### PR TITLE
[ELY-2508] Add tests for DigestMechanismFactory

### PR DESCRIPTION
**Pull request by keshav-redhat**
https://issues.redhat.com/browse/ELY-2508

### CODE REVIEW

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            // OK - expected exception state
+        }
+    }
+
+    @Test
+    public void testCreateAuthenticationMechanismWrongMechanismName() throws HttpAuthenticationException{
+        HttpServerAuthenticationMechanism httpServerAuthenticationMechanism = digestMechanismFactory.createAuthenticationMechanism("MECHANISM_NAME_DOES_NOT_EXISTS",new HashMap<String,String>(),dummyCallbackHandler);
+        Assert.assertNull("Provided mechanism must be null.", httpServerAuthenticationMechanism);
+    }
+}
```
Comment (skyllarr) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java:142]: Hi @keshav-redhat , the PR looks good! Just a minor, there is this task in the issue description that is still missing:


>Also add a test case to both which tests that if different challenge prefix gets used (so eg. try to use BasicMechanismFactory to create authentication mechanism with name "DIGEST" and vice versa), the factory returns null. But with appropriate prefix the factory will create digest / basic authentication mechanism as expected.


What I meant by that is, that it would be nice to add a test here that uses prefix `BASIC` and make sure the `DigestMechanismFactory` returns null in such a case as it dos not provide such a mechanism. 

Another test that could be added here is the case where we have a correct mechanism name, non-empty properties and existing callback and make sure that the resulting HttpServerAuthenticationMechanism is not null.

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    @Test
+    public void testCreateAuthenticationMechanismBasicMechanismName() throws HttpAuthenticationException{
+        HttpServerAuthenticationMechanism httpServerAuthenticationMechanism = digestMechanismFactory.createAuthenticationMechanism("BASIC",new HashMap<String,String>(),dummyCallbackHandler);
+        Assert.assertNull("Provided mechanism must be null.", httpServerAuthenticationMechanism);
+    }
+
+    @Test
+    public void testCreateAuthenticationMechanismReturnNotNull() throws HttpAuthenticationException{
+        HttpServerAuthenticationMechanism httpServerAuthenticationMechanism = digestMechanismFactory.createAuthenticationMechanism("DIGEST",new HashMap<String,String>(),dummyCallbackHandler);
+        Assert.assertNotNull("HttpServerAuthenticationMechanism is not null.",httpServerAuthenticationMechanism);
```
Comment (skyllarr) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: @keshav-redhat Just a detail. the message "HttpServerAuthenticationMechanism is not null." is being used to inform us when the assertion failed. In this case, if the assertion fails it means that the `httpServerAuthenticationMechanism` is null. So please update to:

`"HttpServerAuthenticationMechanism cannot be null."`

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import org.wildfly.security.http.HttpAuthenticationException;
+import org.wildfly.security.http.HttpServerAuthenticationMechanism;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import java.io.IOException;
+import java.util.HashMap;
+
+public class DigestMechanismFactoryTest {
```
Comment (skyllarr) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java:38]: @keshav-redhat Just a minor, can you please add javadoc here? Thank you!

Comment (skyllarr) [reply http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java:38]: At least on top of the class

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    @Test
+    public void testGetMechanismNamesPropertiesNull(){
+        digestMechanismFactory.getMechanismNames(null);
+        digestMechanismFactory.getMechanismNames(new HashMap<String,String>());
+    }
+
+    @Test
+    public void testGetMechanismNamesReturnNotNull(){
+        String[] mechanismNames = digestMechanismFactory.getMechanismNames(null);
+        Assert.assertNotNull("Array of mechanism names is not null.", mechanismNames);
```
Comment (skyllarr) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: s/"Array of mechanism names is not null."/"Array of mechanism names cannot be null."

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
+package org.wildfly.security.http.digest;
```
Comment (skyllarr) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java:19]: @keshav-redhat just a minor, please add the copyright message on top of each new class. Example is [here](https://github.com/wildfly-security/wildfly-elytron/blob/1.x/http/oidc/src/test/java/org/wildfly/security/http/oidc/OidcBaseTest.java#L1)

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2017 Red Hat, Inc., and individual contributors
```
Comment (skyllarr) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: @keshav-redhat Just a minor, year can be updated here to 2023

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java (hunk 4 above)
Comment (skyllarr) [reply http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: @keshav-redhat This was not updated

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java (hunk 2 above)
Comment (skyllarr) [reply http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: @keshav-redhat This was not updated

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java (hunk 1 above)
Comment (skyllarr) [reply http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java:142]: Hello @keshav-redhat  , this was not updated

EDIT: Sorry, I see that the `testCreateAuthenticationMechanismBasicMechanismName` and `testCreateAuthenticationMechanismReturnNotNull` were added and address my comment, so ignor emy comment

**Review by skyllarr — COMMENTED** [review]
Hello @keshav-redhat , I took another look.  Thanks for adding of javadoc and copyright. Other comments were not addressed, maybe you did not add them to the commit?

**Review by skyllarr — APPROVED** [review]

**Comment by skyllarr** [comment]
@keshav-redhat LGTM, thank you!

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        }
+    };
+
+    /**
+     * Tests that getMechanismNames(Map) correctly
+     * handles null or empty properties parameter as possible value.
+     */
+    @Test
+    public void testGetMechanismNamesPropertiesNull(){
+        digestMechanismFactory.getMechanismNames(null);
```
Comment (fjuma) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: Would be good to check the return value here and below.

Comment (keshav-redhat) [reply http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: Thank you for reviewing the PR @fjuma .I have updated the PR.Please have a look.

**Comment by fjuma** [comment]
@keshav-redhat When you get a chance, please update the base branch for this one to 2.x. Thanks!

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        }
+    };
+
+    /**
+     * Tests that getMechanismNames(Map) correctly
+     * handles null or empty properties parameter as possible value.
+     */
+    @Test
+    public void testGetMechanismNamesPropertiesNull(){
+        String []mechanismNames1 = digestMechanismFactory.getMechanismNames(null);
```
Comment (fjuma) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: `String[]`

File: http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        Assert.assertNotNull("Array of mechanism names cannot be null.",mechanismNames1);
+
+        String []mechanismNames2 = digestMechanismFactory.getMechanismNames(new HashMap<String,String>());
```
Comment (fjuma) [inline http/digest/src/test/java/org/wildfly/security/http/digest/DigestMechanismFactoryTest.java]: `String[]`

**Review by fjuma — CHANGES_REQUESTED** [review]
@keshav-redhat Just a couple very minor things and then this should be ready to merge. Thanks!

**Review by fjuma — APPROVED** [review]
Thanks @keshav-redhat!
